provider/datagen/src/lib.rs

```diff
@@ -271,93 +430,78 @@ impl DatagenProvider {
                         .and_then(DataResponse::take_payload)
                         .map_err(|e| e.with_req(key, Default::default()))?;
 
-                    return exporter.flush_singleton(key, &payload).map_err(|e| e.with_req(key, Default::default()));
+                    return exporter
+                        .flush_singleton(key, &payload)
+                        .map_err(|e| e.with_req(key, Default::default()));
                 }
 
-                let mut supported_locales: HashSet<DataLocale> = provider
-                    .supported_locales_for_key(key)
-                    .map_err(|e| e.with_key(key))?.into_iter().collect();
-
-                match provider.source.options.fallback {
-                    options::FallbackMode::Legacy => {
-                        supported_locales
-                            .into_par_iter()
-                            .try_for_each(|locale| {
-                                log::trace!("Generating for key/locale: {key} {locale:?}");
-                                let req = DataRequest {
-                                    locale: &locale,
-                                    metadata: Default::default(),
-                                };
-                                let payload = provider
-                                    .load_data(key, req)
-                                    .and_then(DataResponse::take_payload)
-                                    .map_err(|e| e.with_req(key, req))?;
-                                exporter.put_payload(key, &locale, &payload)
-                            })
-                            .map_err(|e| e.with_key(key))?;
-
-                        exporter
-                            .flush_with_fallback(key, icu_provider::datagen::FallbackMode::None)
-                            .map_err(|e| e.with_key(key))
+                let locales_to_export = provider.select_locales_for_key(key)?;
+
+                match (
+                    provider.source.options.fallback,
+                    exporter.supports_built_in_fallback(),
+                ) {
+                    (options::FallbackMode::Runtime, _)
+                    | (options::FallbackMode::RuntimeManual, _)
+                    | (options::FallbackMode::PreferredForExporter, true) => {
+                        let payloads =
+                            FrozenMap::<DataLocale, Box<DataPayload<ExportMarker>>>::new();
+                        locales_to_export.into_par_iter().try_for_each(|locale| {
+                            let payload = provider.load_with_fallback(key, &locale)?;
+                            if let Some(payload) = payload {
+                                payloads.insert(locale, Box::new(payload));
+                            }
+                            Ok::<(), DataError>(())
+                        })?;
+                        let payloads = payloads
+                            .into_tuple_vec()
+                            .into_iter()
```

Comment:
Afaik with `rayon` you can directly collect into a `Result<Vec<(K, V)>, DataError>>`, no need for `FrozenMap`.

Reply:
Good suggestion. It appears there are 2 ways to do this

1. Map and Collect (f6e5b039ef2ead066e2d99ed4341c427ac2b72c7)
2. Fold and Reduce (eb29a588dac241d889d3091fdf48ec115038360b)

I think (2) is supposed to be more efficient because (1) requires rayon to keep track of the relative order of the elements across all the threads, collect each one of them, and then flush them into the collector all at the end, whereas (2) folds each processor's results into a single collection, and then merges those collections at the end. However, I think an advantage of (1) would be that we get a single nice error message when one of the jobs fails, instead of the multi-core mess of an error stream that we currently get.

Happy to do either one.

Reply:
I think (1) as you currently have is simpler and I don't think there'll be a big performance impact (I hope rayon doesn't keep all threads alive).